// bundle-buddy/graph_builder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
function sumLines(obj) {
    return Object.keys(obj).map(k => obj[k].count).reduce((p, c) => {
        return p + c;
    }, 0);
}
function buildLinks(groupedOutputFiles, logger) {
    logger.info("Building graph links between bundles.");
    const ret = [];
    for (const bundleName of groupedOutputFiles.keys()) {
        const sourceFiles = Object.keys(groupedOutputFiles.get(bundleName));
        for (const sourcePath of sourceFiles) {
            ret.push({
                source: sourcePath,
                target: bundleName,
                strength: groupedOutputFiles.get(bundleName)[sourcePath].count
            });
        }
    }
    return ret;
}
function buildGraph(groupedSourceFiles, groupedOutputFiles, sourceToBundles, logger) {
    const nodes = [];
    logger.info("Building graph between bundles.");
    // build input nodes
    for (const sourceFileName of groupedSourceFiles.keys()) {
        nodes.push({
            id: sourceFileName,
            type: "input",
            size: sumLines(groupedSourceFiles.get(sourceFileName)),
            inBundleFiles: Array.from(sourceToBundles[sourceFileName])
        });
    }
    // build output nodes
    for (const outputFileName of groupedOutputFiles.keys()) {
        nodes.push({
            id: outputFileName,
            type: "output",
            size: sumLines(groupedOutputFiles.get(outputFileName))
        });
    }
    return {
        nodes,
        links: buildLinks(groupedOutputFiles, logger)
    };
}
exports.buildGraph = buildGraph;
