// bundle-buddy/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const chalk = require("chalk");
const path = require("path");
const WEBPACK_MATCHER = /\/\/ WEBPACK FOOTER \/\/\n\/\/\s+(.*)/m;
const HASH_SPLITTER = "||||";
exports.VIZ_PATH = "viz/build";
function hashFileLineNumber(fileName, lineNumber) {
    return `${fileName}${HASH_SPLITTER}${lineNumber}`;
}
exports.hashFileLineNumber = hashFileLineNumber;
function hashToFileAndLineNumber(hash) {
    return {
        fileName: hash.split(HASH_SPLITTER)[0],
        lineNumber: parseInt(hash.split(HASH_SPLITTER)[1], 10)
    };
}
exports.hashToFileAndLineNumber = hashToFileAndLineNumber;
function hashBundlesToKey(files) {
    return Array.from(files).sort().join(HASH_SPLITTER);
}
exports.hashBundlesToKey = hashBundlesToKey;
function formatProcessedSourceMaps(processed) {
    return JSON.stringify({
        graph: processed.graph,
        sourceFiles: processed.sourceFiles,
        bundleFileStats: [...processed.bundleFileStats],
        outputFiles: processed.outputFiles,
        groupedBundleStats: [...processed.groupedBundleStats],
        perFileStats: [...processed.perFileStats],
        sourceFileLinesGroupedByCommonBundle: processed.sourceFileLinesGroupedByCommonBundle
    });
}
exports.formatProcessedSourceMaps = formatProcessedSourceMaps;
/**
 * Since the sourcemap file name does not always === the real file name
 * we need to do some cleanup work (specifically for webpack).
 */
function getOriginalFileNameFromSourceName(sourceName, sourceMapConsumer) {
    const contents = sourceMapConsumer.sourceContentFor(sourceName);
    // https://twitter.com/samccone/status/878773452169027588
    const match = WEBPACK_MATCHER.exec(contents);
    if (match && match[1]) {
        return match[1];
    }
    return sourceName;
}
exports.getOriginalFileNameFromSourceName = getOriginalFileNameFromSourceName;
function sourceMapToLineHits(hitTracks) {
    const sourceToLineMapping = new Set();
    for (const fileName of hitTracks.keys()) {
        Object.keys(hitTracks.get(fileName).inUse).forEach(lineNumber => {
            sourceToLineMapping.add(hashFileLineNumber(fileName, lineNumber));
        });
    }
    return sourceToLineMapping;
}
exports.sourceMapToLineHits = sourceMapToLineHits;
/**
 * @export
 * @param {string} dataPath Name of file being created.
 * @param {string} [contextPath=__dirname] Path to be passed in when function is not consumed by bundle-buddy cli process
 * @returns {string} Absolute path where data will be written to
 */
function getWritePathForSerializedData(dataPath, contextPath = __dirname) {
    return path.join(contextPath, exports.VIZ_PATH, dataPath || `data_${Date.now()}`);
}
exports.getWritePathForSerializedData = getWritePathForSerializedData;
class Logger {
    constructor(opts) {
        this.logLevel = opts.logLevel;
    }
    error(s) {
        console.warn(chalk.white.bgRed.bold(s));
    }
    success(s) {
        console.warn(chalk.white.bgGreen.bold(s));
    }
    info(s) {
        if (this.logLevel === "verbose") {
            console.warn(chalk.grey(s));
        }
    }
}
exports.Logger = Logger;
